hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/stack/savepoints/AbstractSavepoint.java

```diff
@@ -0,0 +1,161 @@
+ * <p>
+ * When adopting block streams we will add more extension points to this class in the form of abstract methods
+ * that determine how each type of savepoint constructs state change block items.
+ */
+public abstract class AbstractSavepoint extends BuilderSinkImpl implements Savepoint {
+    private static final EnumSet<ResponseCodeEnum> SUCCESSES =
+            EnumSet.of(OK, SUCCESS, FEE_SCHEDULE_FILE_PART_UPLOADED, SUCCESS_BUT_MISSING_EXPECTED_OPERATION);
+
+    protected final BuilderSink parentSink;
+    protected final WrappedHederaState state;
+    private Status status = Status.PENDING;
+
+    /**
+     * Constructs a savepoint with limits that discriminate between the number of preceding and following builders.
+     * @param state the current state
+     * @param parentSink the parent sink
+     * @param maxPreceding the maximum number of preceding builders
+     * @param maxFollowing the maximum number of following builders
+     */
+    protected AbstractSavepoint(
+            @NonNull final WrappedHederaState state,
+            @NonNull final BuilderSink parentSink,
+            final int maxPreceding,
+            final int maxFollowing) {
+        super(maxPreceding, maxFollowing);
+        this.state = requireNonNull(state);
+        this.parentSink = requireNonNull(parentSink);
+    }
+
+    /**
+     * Constructs a savepoint with a total limit on the number of builders that can be accumulated.
+     * @param state the current state
+     * @param parentSink the parent sink
+     * @param maxTotal the maximum number of total builders
+     */
+    protected AbstractSavepoint(
+            @NonNull final WrappedHederaState state, @NonNull final BuilderSink parentSink, final int maxTotal) {
+        super(maxTotal);
+        this.state = requireNonNull(state);
+        this.parentSink = requireNonNull(parentSink);
+    }
+
+    @Override
+    public HederaState state() {
+        return state;
+    }
+
+    @Override
+    public void commit() {
+        assertNotFinished();
+
+        state.commit();
+        commitBuilders();
+        status = Status.FINISHED;
+    }
+
+    @Override
+    public void rollback() {
+        assertNotFinished();
+
+        rollback(precedingBuilders);
+        rollback(followingBuilders);
+        commitBuilders();
+        status = Status.FINISHED;
+    }
+
+    @Override
+    public SingleTransactionRecordBuilder createBuilder(
+            @NonNull final SingleTransactionRecordBuilder.ReversingBehavior reversingBehavior,
+            @NonNull final HandleContext.TransactionCategory txnCategory,
+            @NonNull final ExternalizedRecordCustomizer customizer,
+            final boolean isBaseBuilder) {
+        requireNonNull(reversingBehavior);
+        requireNonNull(txnCategory);
+        requireNonNull(customizer);
+        final var builder = new SingleTransactionRecordBuilderImpl(reversingBehavior, customizer, txnCategory);
+        if (!customizer.shouldSuppressRecord()) {
```

Comment:
What use cases do we have for suppressing a record?

Reply:
It is used in `ContractService`  [here](https://github.com/hashgraph/hedera-services/blob/13889-Savepoint/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/scope/HandleHederaOperations.java#L372)

> // Create should have conditional child record, but we only externalize this child if it's not already
        // externalized by the top-level HAPI transaction; and we "finish" the synthetic transaction by swapping
        // in the contract creation body for the dispatched crypto create body

Reply:
Great, thanks for the example :)